### PR TITLE
Convert `StringAgg` to UDAF

### PR DESCRIPTION
**Pull request by lewiszlw**
## Which issue does this PR close?

part of https://github.com/apache/datafusion/issues/8708.

### CODE REVIEW

File: datafusion/functions-aggregate/src/string_agg.rs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+            }
+            _ => not_impl_err!(
+                "StringAgg not supported for delimiter {}",
+                &acc_args.input_exprs[1]
+            ),
+        }
+    }
+
+    fn aliases(&self) -> &[String] {
+        &self.aliases
```
Comment (jayzhan211) [inline datafusion/functions-aggregate/src/string_agg.rs:105]: we can return `&[]` here

File: datafusion/proto/tests/cases/roundtrip_physical_plan.rs [hunk 2]
```diff
@@ -356,13 +356,6 @@ fn rountrip_aggregate() -> Result<()> {
             Vec::new(),
             Vec::new(),
         ))],
-        // STRING_AGG
-        vec![Arc::new(StringAgg::new(
```
Comment (jayzhan211) [inline datafusion/proto/tests/cases/roundtrip_physical_plan.rs:360]: I think we should create expression with `udaf::create_aggregate_expr`

**Review by jayzhan211 — COMMENTED** [review]
Also, remember to add test in `roundtrip_expr_api`

File: datafusion/functions-aggregate/src/string_agg.rs (hunk 1 above)
Comment (lewiszlw) [reply datafusion/functions-aggregate/src/string_agg.rs:105]: I kept `aliases` field to let users can create aliases if they need.

Comment (jayzhan211) [reply datafusion/functions-aggregate/src/string_agg.rs:105]: I think they can just take a look on other functions that have aliases 😅

Comment (lewiszlw) [reply datafusion/functions-aggregate/src/string_agg.rs:105]: `AliasedAggregateUDFImpl` can be used in this case.

File: datafusion/proto/tests/cases/roundtrip_physical_plan.rs (hunk 2 above)
Comment (lewiszlw) [reply datafusion/proto/tests/cases/roundtrip_physical_plan.rs:360]: Added.

**Review by jayzhan211 — APPROVED** [review]
👍

**Review by alamb — APPROVED** [review]
Thank you @lewiszlw  and @jayzhan211

**Pull request merged**